src/ophys_etl/modules/segmentation/filter/roi_filter.py

```diff
@@ -0,0 +1,253 @@
+from typing import List, Dict, Optional, Union
+from abc import ABC, abstractmethod
+import argschema
+import json
+import h5py
+import pathlib
+import numpy as np
+
+from ophys_etl.modules.decrosstalk.ophys_plane import OphysROI
+
+from ophys_etl.modules.segmentation.merge.roi_utils import (
+    ophys_roi_to_extract_roi)
+
+from ophys_etl.modules.segmentation.qc_utils.roi_comparison_utils import (
+    roi_list_from_file)
+
+
+class ROIBaseFilter(ABC):
+    """
+    Abstract base class for ROI filtering mechanisms.
+
+    Sub-classes of this calss need to implement
+    self.is_roi_valid which accepts an OphysROI and returns
+    a boolean indicating whether or not that ROI is accepted
+    as valid according to the filter's criterion.
+
+    The __init__ for the sub-class also needs to define
+    self._reason, a string summarizing what the filter does
+    so that users can know why their ROIs were flagged as
+    invalid.
+    """
+
+    _reason = None  # reason ROI is marked as invalid
+
+    @abstractmethod
+    def is_roi_valid(self, roi: OphysROI) -> bool:
+        """
+        Determines if an ROI passes the filter
+
+        Parameters
+        ----------
+        roi: OphysROI
+
+        Returns
+        -------
+        bool
+            True if the ROI is determined as valid;
+            False otherwise
+        """
+        raise NotImplementedError()
+
+    def do_filtering(
+            self,
+            roi_list: List[OphysROI]) -> Dict[str, List[OphysROI]]:
+        """
+        Loop over a list of ROIs, testing each agains the filter's
+        criterion. Returns a dict in which 'valid_roi' points to
+        a list of ROIs that passed the cut and 'invalid_roi' points
+        to a list of ROIs that did not. Invalid ROIs will be
+        re-instantiated with valid_roi=False
+
+        Parameters
+        ----------
+        roi_list: List[OphysROI]
+
+        Returns
+        -------
+        Dict[str, List[OphysROI]]
+        """
+        valid_roi = []
+        invalid_roi = []
+        for roi in roi_list:
+            if self.is_roi_valid(roi):
+                valid_roi.append(roi)
+            else:
+                new_roi = OphysROI(
```

Comment:
I don't like the idea of `do_filtering` changing its inputs in-place. I know that the way this method is currently used, the list of ROIs it is passed only exists within the `FilterRunner`, but I don't want users to be surprised if they try to use this in a non-traditional way.

I just merged a commit so that `valid_roi` is populated with copies of the original ROIs so that there are no "pass by reference" shenanigans going on.

Reply:
What I was sort of worried about was the verbosity of instantiating a new roi. What about
```
new_roi = copy.deepcopy(roi)
new_roi.valid_roi = False
```